docgen/parser/parser.go

```diff
@@ -0,0 +1,213 @@
+			Comment: comments.Get(fmt.Sprintf("%s.%d.%d", path, fieldFlag, i)),
+			Type:    getType(f),
+		}
+	}
+	return res
+}
+
+func getType(f *google_protobuf.FieldDescriptorProto) string {
+	if t := f.GetTypeName(); t != "" {
+		s := strings.Split(t, ".")
+		return s[len(s)-1]
+	}
+	return strings.ToLower(strings.Replace(f.GetType().String(), "TYPE_", "", 1))
+}
+
+func parseServices(pkgName string, messages map[string]*Message, services []*google_protobuf.ServiceDescriptorProto) (map[string]*Service, error) {
+	res := map[string]*Service{}
+	for _, s := range services {
+		methods, err := parseMethods(pkgName, messages, s.GetMethod())
+		if err != nil {
+			return nil, err
+		}
+		res[getQualifiedName(pkgName, s.GetName())] = &Service{
+			Name:    s.GetName(),
+			Methods: methods,
+		}
+	}
+	return res, nil
+}
+
+func parseMethods(pkgName string, messages map[string]*Message, methods []*google_protobuf.MethodDescriptorProto) (map[string]*Method, error) {
+	res := map[string]*Method{}
+	for _, m := range methods {
+		extOp, err := proto.GetExtension(m.GetOptions(), options.E_Openapiv2Operation)
+		if err != nil {
+			if err == proto.ErrMissingExtension {
+				continue
+			}
+			return nil, err
+		}
+
+		extHTTP, err := proto.GetExtension(m.GetOptions(), method.E_Http)
+		if err != nil {
+			if err == proto.ErrMissingExtension {
+				continue
+			}
+			return nil, err
+		}
+
+		req, err := parseRequest(extHTTP.(*method.HttpRule), messages[m.GetInputType()])
+		if err != nil {
+			return nil, err
+		}
+
+		res[getQualifiedName(pkgName, m.GetName())] = &Method{
+			Name:      m.GetName(),
+			Response:  messages[m.GetOutputType()],
+			Operation: extOp.(*options.Operation),
+			Request:   req,
+		}
+	}
+	return res, nil
+}
+
+func parseRequest(rule *method.HttpRule, message *Message) (*Request, error) {
+	var verb, uri string
+
+	switch p := rule.GetPattern().(type) {
+	case *method.HttpRule_Get:
+		verb = "GET"
+		uri = p.Get
+	case *method.HttpRule_Post:
+		verb = "POST"
+		uri = p.Post
+	case *method.HttpRule_Put:
+		verb = "PUT"
+		uri = p.Put
+	case *method.HttpRule_Delete:
+		verb = "DELETE"
+		uri = p.Delete
+	default:
+		return nil, fmt.Errorf("%t not supported", p)
+	}
+
+	var body *Message
+	if rule.GetBody() != "" {
+		body = message
+	}
+
+	query, err := parseQuery(uri, message)
+	if err != nil {
+		return nil, err
+	}
+	return &Request{
+		Verb:  verb,
+		URI:   uri,
+		Body:  body,
+		Query: query,
+	}, nil
+}
+
+func parseQuery(uri string, message *Message) ([]*Field, error) {
+	p, err := httprule.Parse(uri)
+	if err != nil {
+		return nil, err
+	}
+
+	tmpl := p.Compile()
+	res := make([]*Field, len(tmpl.Fields))
+	for i, tf := range tmpl.Fields {
+		for _, mf := range message.Fields {
+			if tf == mf.Name {
+				res[i] = mf
+			}
+		}
+	}
+
+	return res, nil
+}
+
+// Get gets a comment by its location path.
+func (c Comments) Get(path string) *Comment {
+	if val, ok := c[path]; ok {
+		return val
+	}
+	return &Comment{}
+}
```

Comment:
I think it would be simpler to remove this Comments type and the method, and use the map directly with a helper. Like:

```
func nonNilComment(c *Comment) *Comment {
    if c == nil { return &Comment{})
    return c
}

[...]

Comment: nonNilComment(comments["str"]),
```

This is less code, the caller side is more intuitive, and there's less getter magic going on.